### PR TITLE
Add support for blob-level KMS encryption keys

### PR DESCRIPTION
**Pull request by tseaver**

### CODE REVIEW

File: storage/google/cloud/storage/blob.py [hunk 1]
```diff
@@ -120,6 +120,12 @@ class Blob(_PropertyMixin):
     :param encryption_key:
         Optional 32 byte encryption key for customer-supplied encryption.
         See https://cloud.google.com/storage/docs/encryption#customer-supplied.
+
+    :type kms_encryption_key: bytes
+    :param kms_encryption_key:
+        Optional esource name of Cloud KMS key used to encrypt the blob's
```
Comment (jonparrott) [inline storage/google/cloud/storage/blob.py]: `resource name`?

File: storage/tests/unit/test_blob.py [hunk 2]
```diff
@@ -459,6 +492,27 @@ def test__get_download_url_on_the_fly_with_user_project(self):
...
+        blob_name = 'bzzz-fly.txt'
+        bucket = _Bucket(name='buhkit')
+        blob = self._make_one(
+            blob_name, bucket=bucket, kms_encryption_key=kms_resource)
+
+        self.assertIsNone(blob.media_link)
+        download_url = blob._get_download_url()
+        expected_url = (
+            'https://www.googleapis.com/download/storage/v1/b/'
+            'buhkit/o/bzzz-fly.txt?alt=media&{}'.format(kms_encoded))
```
Comment (frankyn) [inline storage/tests/unit/test_blob.py]: Why did are you providing kmsKeyName for a GET?

File: storage/google/cloud/storage/blob.py [hunk 3]
```diff
@@ -120,6 +120,12 @@ class Blob(_PropertyMixin):
     :param encryption_key:
         Optional 32 byte encryption key for customer-supplied encryption.
         See https://cloud.google.com/storage/docs/encryption#customer-supplied.
+
+    :type kms_encryption_key: bytes
+    :param kms_encryption_key:
```
[the remark this answers is not in the document]
Comment (tseaver) [reply storage/google/cloud/storage/blob.py]: I agree.  3d728ad

File: storage/tests/unit/test_blob.py (hunk 2 above)
Comment (tseaver) [reply storage/tests/unit/test_blob.py]: I guess I was following the CSEK case too closely.  4464474

File: storage/tests/unit/test_blob.py [hunk 4]
```diff
@@ -459,6 +492,26 @@ def test__get_download_url_on_the_fly_with_user_project(self):
...
+        blob_name = 'bzzz-fly.txt'
+        bucket = _Bucket(name='buhkit')
+        blob = self._make_one(
+            blob_name, bucket=bucket, kms_key_name=kms_resource)
+
+        self.assertIsNone(blob.media_link)
+        download_url = blob._get_download_url()
+        expected_url = (
+            'https://www.googleapis.com/download/storage/v1/b/'
+            'buhkit/o/bzzz-fly.txt?alt=media')  # kmsKeyName *not* expected
```
Comment (frankyn) [inline storage/tests/unit/test_blob.py]: That's correct. KMS key name associated to the object will be used on a GET request. The prerequisites are in the one-pager.

**Comment by jonparrott** [comment]
She's been having some trouble with automerge lately, so let's see if it actually works.

**Comment by tseaver** [comment]
Merging in spite of unrelated [logging systest failures](https://circleci.com/gh/GoogleCloudPlatform/google-cloud-python/6412).

**Pull request merged**